### DISCUSSION TITLE
RefreshEnv doesn't always work from inside batch file?

### DISCUSSION THREAD

**Discussion by B-Manninen** [chocolatey/choco#2428] [Q&A]
I had created some .bat scripts to switch which version of python is on my path (I know there are solutions for this, but I chose to do this in part as an exercise).  The issue of course is that when I edit environment variables I need to refresh them so I can use them in the current terminal window.  RefreshEnv worked great for this - in my proof of concept script.  When I improved my script to automatically detect installed versions of python (so I wouldn't have to hardcode each version) RefreshEnv seems not to work.  This is the working proof of concept - it edits the variables, and then refreshes the terminals environment variables.

```
@echo off
echo "start %PYTHON_HOME%"
set PYTHON_HOME=C:\Python\cs1110-anaconda
setx PYTHON_HOME "%PYTHON_HOME%"
set Path=%PYTHON_HOME%;%Path%
echo "end %PYTHON_HOME%"
echo python 3.8 64 bit has been activated 
RefreshEnv
```

My improved version takes an argument, checks if there is a folder that matches that argument, and if so sets that folder to be the new python home.  It uses the same code to set/refresh, but for some reason it doesn't refresh.  It will edit the variables and print out the "Refreshing environment variables from registry" message, but when I check if they updated they aren't.  If I run RefreshEnv manually they update, so I know the environment variables were edited, just not refreshed.

```
@echo off
SETLOCAL EnableDelayedExpansion

if "%~1"=="" (
    echo No parameters have been provided.

    GOTO :eof
) else (
    for /d %%d in (..\*.*) do (
        set str=%%d
        if "%~1"=="ava" (
            echo !str:~3%!
        ) else (
	    if "%~1"=="!str:~3%!" (
                GOTO :swap
            )
        )
    )
    if NOT "%~1"!=="ava" echo you must pass a valid version of python that is installed on the system
    GOTO :eof
)

:swap
@echo off
echo "start %PYTHON_HOME%"
set PYTHON_HOME="C:\Python\%~1"
setx PYTHON_HOME "%PYTHON_HOME%"
set Path=%PYTHON_HOME%;%Path%
echo "end %PYTHON_HOME%"
echo python %~1 has been activated 
RefreshEnv
```

(edits - originally I messed up the code formatting)

**Comment by B-Manninen** [comment]
I've tried using `call RefreshEnv.cmd`, and calling a different .bat file that just had RefreshEnv, neither helped.

Also here is an example of my usage - I start with the the 'cs1110' version (3.8 anaconda), then I use the old script to swap to the 3.10 version, and when I check it has updated correctly.

```
C:\Python\scripts>echo %PYTHON_HOME%
C:\Python\cs1110-anaconda

C:\Python\scripts>python310
"start C:\Python\cs1110-anaconda"

SUCCESS: Specified value was saved.
"end C:\Python\Python3-10-32x"
Refreshing environment variables from registry for cmd.exe. Please wait...Finished..

C:\Python\scripts>echo %PYTHON_HOME%
C:\Python\Python3-10-32x
```
Then I try the script that takes an argument, and I check to see that it has not updated.
```
C:\Python\scripts>pyswap cs1110-anaconda
"start C:\Python\Python3-10-32x"

SUCCESS: Specified value was saved.
"end "C:\Python\cs1110-anaconda""
python cs1110-anaconda has been activated
Refreshing environment variables from registry for cmd.exe. Please wait...Finished..

C:\Python\scripts>echo %PYTHON_HOME%
C:\Python\Python3-10-32x
```
But if I manually refresh it is clear that the environment variable did update, but RefreshEnv didn't work for some reason
```
C:\Python\scripts>RefreshEnv
Refreshing environment variables from registry for cmd.exe. Please wait...Finished..

C:\Python\scripts>echo %PYTHON_HOME%
C:\Python\cs1110-anaconda

C:\Python\scripts>
```

(edit - formatting again)

**Reply by gep13** [reply]
Can you confirm if you are running within a PowerShell session when running the above?

**Reply by B-Manninen** [reply]
Actually I'm in cmd

**Comment by rasa** [comment]
Perhaps it's not refreshing the environment because you first called `setlocal`?

Try executing `endlocal` before running `refreashenv.cmd`:

```batch
endlocal
RefreshEnv.cmd
```
or
```batch
endlocal
call RefreshEnv.cmd
@rem more commands...

```
